crates/nu-command/src/strings/mime_/guess.rs

```diff
@@ -0,0 +1,160 @@
+                    ],
+                    NO_SPAN,
+                )),
+            },
+            Example {
+                example: r#"let input = glob -d 1 * | wrap filename; $input | merge ($input | get filename | mime guess | select type)"#,
+                description: "Add a MIME type column to a table.",
+                result: Some(Value::list(
+                    vec![Value::record(
+                        record!("filename" => Value::string("...".to_string(), NO_SPAN), "type" => Value::string("...", NO_SPAN)),
+                        NO_SPAN,
+                    )],
+                    NO_SPAN,
+                )),
+            },
+        ]
+    }
+
+    fn run(
+        &self,
+        engine_state: &EngineState,
+        stack: &mut Stack,
+        call: &Call,
+        input: PipelineData,
+    ) -> Result<PipelineData, ShellError> {
+        let use_extension: bool = call.has_flag(engine_state, stack, "extension")?;
+
+        let guess_function: fn(&str) -> mime_guess::MimeGuess = if use_extension {
+            mime_guess::from_ext
+        } else {
+            // HACK I don't know how to satisfy the compiler here without a closure, but I cannot return the function directly.
+            // If I do, I get an error that the types are different or that a value does not live long enough when the function is called.
+            |input| mime_guess::from_path(input)
+        };
+
+        match input {
+            PipelineData::Value(Value::String { val, internal_span }, ..) => {
+                let mime_type = guess_function(&val)
+                    .first()
+                    .map(|mime| mime.to_string())
+                    .unwrap_or_else(|| "unknown".to_string());
+
+                Ok(Value::string(mime_type, internal_span).into_pipeline_data())
+            }
+            PipelineData::Value(Value::List { .. }, ..) | PipelineData::ListStream(..) => {
+                let mime_records_iter = input.into_iter().map(move |value| {
+                    let span = value.span();
+
+                    match value.as_str() {
+                        Ok(s) => {
+                            let name = Value::string(s, span);
+                            let mime_type = Value::string(
+                                guess_function(s)
+                                    .first()
+                                    .map(|mime| mime.to_string())
+                                    .unwrap_or_else(|| "unknown".to_string()),
+                                span,
+                            );
+
+                            Value::record(record!("name" => name, "type" => mime_type), span)
+                        }
+                        Err(err) => Value::error(
+                            ShellError::TypeMismatch {
+                                err_message: err.to_string(),
+                                span,
+                            },
+                            span,
+                        ),
```

Comment:
`to_string`ing the whole original error seems wrong. Either simply keep the `ShellError::CantConvert` from `Value::as_str` or do a proper `ShellError::OnlySupportsThisInputType`